docs/migration-guide/from-older-version.md

```diff
@@ -0,0 +1,158 @@
+
+### Upgrade MDX1 to MDX3
+
+Storybook 8 uses MDX3 by default for rendering [docs](./writing-docs/index.md). The upgrade from MDX1 to MDX3 is not fully automated due to the large number of changes between versions (particularly v1 to v2). Fortunately, we have some tips to help make it a lot easier.
+
+#### Automatically detect MDX2 errors with a CLI tool
+
+If your project contains MDX files, run the following command before starting up Storybook:
+
+```sh
+npx @hipster/mdx2-issue-checker
+```
+
+This will go through every MDX file in the current working directory, and show you which files have errors:
+
+![Terminal output of mdx2-issue-checker, showing errors found](./assets/mdx2-issue-checker-errors.png)
+
+#### Fix MDX2 errors iteratively
+
+The CLI only shows the first error per file, so you might need to run the checker iteratively. One way to streamline this process if you use VS Code is to:
+
+1. Install the [MDX VS Code extension](https://marketplace.visualstudio.com/items?itemName=unifiedjs.vscode-mdx)
+2. Enable experimental IntelliSense support for MDX files in your user settings: `"mdx.experimentalLanguageServer": true`
+
+This shows the errors visually in your editor, which speeds things up a lot. Here's what it looks like to fix multiple errors in a file using the extension:
+
+![MDX errors showing in VS Code](./assets/mdx-vs-code-extension-errors.gif)
+
+### `storiesOf` to CSF
+
+Storybook architecture focuses on performance and now needs statically analyzable code. For that reason, it does not work with `storiesOf`. We provide a codemod which, in most cases, should automatically make the code changes for you (make sure to update the glob to fit your files):
+
+<!-- prettier-ignore-start -->
+
+<CodeSnippets
+  paths={[
+    'common/storybook-migrate-stories-of-to-csf.npm.js.mdx',
+    'common/storybook-migrate-stories-of-to-csf.pnpm.js.mdx',
+    'common/storybook-migrate-stories-of-to-csf.yarn.js.mdx'
+  ]}
+/>
+
+<!-- prettier-ignore-end -->
+
+This will transform your stories into [CSF 1](/blog/component-story-format/) stories, which are story functions that don't accept [args](../writing-stories/args.md). These are fully supported in Storybook 8, and will continue to be for the foreseeable future.
+
+However, we recommend [writing all **new** stories in CSF 3](../writing-stories/index.md), which are story objects that are easier to write, reuse, and maintain.
+
+### `storiesOf` to dynamically created stories
+
+If you are using `storiesOf` for its ability to create stories dynamically, you can build your own "storiesOf" implementation by leveraging the new [(experimental) indexer API](../api/main-config-indexers). A proof of concept (with a deeper explanation of the implementation) can be seen in [this StackBlitz demo](https://stackblitz.com/edit/github-h2rgfk?file=README.md).
```

Comment:
If we detect that the user uses `storyStoryV7: false`, we block the upgrade to 8. And the `storiesOf` API ONLY works with `storyStoryV6 (storyStoryV7: false)`. So in theory, the user has to remove `storyStoryV7: false` to allow the upgrade. In this middle state, Storybook will be broken for user's who continue to use `storiesOf` in Storybook 6. So they upgrade to Storybook 8 and then they have to run the necessary steps to get rid of the `storiesOf` API. I just wanted to clarify this.